pkg/infrastructure/azure/storage.go

```diff
@@ -206,30 +254,83 @@ func CreatePageBlob(ctx context.Context, in *CreatePageBlobInput) (*CreatePageBl
 		},
 	)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get page blob client: %w", err)
+		return "", fmt.Errorf("failed to get page blob client: %w", err)
 	}
 
-	// This is used in terraform, not sure if it matters
-	metadata := make(map[string]*string, 1)
-	metadata["source_uri"] = to.Ptr(in.ImageURL)
+	logrus.Debugf("Creating Page blob and uploading image to it")
+	if in.ImageURL == "" {
+		_, err = pageBlobClient.Create(ctx, in.ImageLength, nil)
+		if err != nil {
+			return "", fmt.Errorf("failed to create page blob with image contents: %w", err)
+		}
+		// This image (example: ignition shim) needs to be uploaded from a local file.
+		err = doUploadPages(ctx, pageBlobClient, in.BootstrapIgnData, in.ImageLength)
+		if err != nil {
+			return "", fmt.Errorf("failed to upload page blob image contents: %w", err)
+		}
+	} else {
+		// This is used in terraform, not sure if it matters
+		metadata := make(map[string]*string, 1)
+		metadata["source_uri"] = to.Ptr(in.ImageURL)
+
+		_, err = pageBlobClient.Create(ctx, in.ImageLength, &pageblob.CreateOptions{
+			Metadata: metadata,
+		})
+		if err != nil {
+			return "", fmt.Errorf("failed to create page blob with image URL: %w", err)
+		}
 
-	logrus.Debugf("Creating blob")
-	_, err = pageBlobClient.Create(ctx, in.ImageLength, &pageblob.CreateOptions{
-		Metadata: metadata,
-	})
-	if err != nil {
-		return nil, fmt.Errorf("failed to create blob: %w", err)
+		err = doUploadPagesFromURL(ctx, pageBlobClient, in.ImageURL, in.ImageLength)
+		if err != nil {
+			return "", fmt.Errorf("failed to upload page blob image from URL %s: %w", in.ImageURL, err)
+		}
 	}
 
-	logrus.Debugf("Uploading to blob")
-	err = doUploadPagesFromURL(ctx, pageBlobClient, in.ImageURL, in.ImageLength)
+	// Is this addition OK for when CreatePageBlob() is called from InfraReady()
+	sasURL, err := pageBlobClient.GetSASURL(sas.BlobPermissions{Read: true}, time.Now().Add(time.Minute*60), &blob.GetSASURLOptions{})
 	if err != nil {
-		return nil, fmt.Errorf("failed to upload blob image %s: %w", in.ImageURL, err)
+		return "", fmt.Errorf("failed to get Page Blob SAS URL: %w", err)
 	}
-	return &CreatePageBlobOutput{
-		PageBlobClient:      pageBlobClient,
-		SharedKeyCredential: sharedKeyCredential,
-	}, nil
+	return sasURL, nil
+}
+
+func doUploadPages(ctx context.Context, pageBlobClient *pageblob.Client, imageData []byte, imageLength int64) error {
+	logrus.Debugf("Uploading to Page Blob with Image of length :%d", imageLength)
+
+	// Padding Image to align with 512 boundary
+	logrus.Debugf("Original Image length: %d", int64(len(imageData)))
+	padding := imageLength - int64(len(imageData))
+	paddingString := strings.Repeat(" ", int(padding)) + string(imageData[len(imageData)-1])
+	imageData = append(imageData[0:len(imageData)-1], paddingString...)
```

Comment:
Can you add a comment explaining why it's ok to add the padding just before the last character of `imageData`?